runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,238 @@
+                    .cacheEntries(!runtime.getParameters().isCacheDisabled())
+                    .build();
+
+            this.txnStream = runtime.getStreamsView()
+                    .get(ObjectsView.TRANSACTION_STREAM_ID, options);
+            this.txnStream.seek(startAddress + 1);
+        }
+
+        public void run() {
+            ILogData logData = null;
+            do {
+                try {
+                    Thread.currentThread().setName(namespace + listener.toString());
+                    logData = txnStream.nextUpTo(Address.MAX);
+                    if (logData == null) {
+                        break; // Stream is all caught up, sleep for a bit.
+                    }
+                    MultiObjectSMREntry multiObjSMREntry = (MultiObjectSMREntry) logData.getPayload(runtime);
+                    long epoch = logData.getEpoch();
+                    Map<TableSchema, List<CorfuStreamEntry>> entries = new HashMap<>();
+                    // first only filter by the stream IDs of interest that are present in this logData
+                    logData.getStreams().stream().filter(tablesOfInterest::containsKey)
+                            .forEach(streamId -> entries.put(tablesOfInterest.get(streamId),
+                                    // Only extract the list of updates per stream as a list
+                                    multiObjSMREntry.getSMRUpdates(streamId).stream().map(smrEntry ->
+                                            CorfuStreamEntry.fromSMREntry(smrEntry,
+                                                    epoch,
+                                                    tablesOfInterest.get(streamId).getKeyClass(),
+                                                    tablesOfInterest.get(streamId).getPayloadClass(),
+                                                    tablesOfInterest.get(streamId).getMetadataClass())
+                                    ).collect(Collectors.toList())));
+
+                    if (!entries.isEmpty()) {
+                        CorfuStreamEntries callbackResult = new CorfuStreamEntries(entries);
+                        log.debug("{}::onNext with {} updates", listener.toString(), entries.size());
+                        long onNextStart = System.nanoTime();
+                        listener.onNext(callbackResult);
+                        long onNextEnd = System.nanoTime();
+                        if (TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart) > 0) {
+                            log.debug("{}::onNext took {}s", listener.toString(),
+                                    TimeUnit.NANOSECONDS.toSeconds(onNextEnd - onNextStart));
+                        }
+                    }
+                } catch (Throwable throwable) {
+                    log.warn("{}::onError: {}", listener.toString(), throwable.toString());
+                    listener.onError(throwable);
+                    streamManager.unsubscribe(listener);
```

Comment:
hmm unsubscribe() calls the executor pool's thread cancel will might interrupt the thread. I don't know the implications of executing code after interrupt is invoked on the thread.

Reply:
This is another comment I had, but that was folded by github, so you probably didn't see it, so pasting here:

"I suggest not pass true to cancel(). This will raise UnrecoverableCorfuInterruptedError. We saw a lot of such bugs, and nowadays we are confident to punt these bugs to verticals because corfu never do thread.interupt() or future.cancel(true)."

Reply:
now fixed based on offline discussion about not interrupting thread at all. Brought back the shutdown variable.